packages/loopring_v3/contracts/impl/BasicDepositContract.sol

```diff
@@ -3,66 +3,77 @@
 pragma solidity ^0.6.10;
 
 import "../iface/IDepositContract.sol";
-import "../iface/ILoopringV3.sol";
 
 import "../lib/AddressUtil.sol";
+import "../lib/Claimable.sol";
 import "../lib/ERC20.sol";
 import "../lib/ERC20SafeTransfer.sol";
 import "../lib/MathUint.sol";
 import "../lib/ReentrancyGuard.sol";
 
 
-/// @title Basic implementation of IDepositContract that just stores
+/// @title Default implementation of IDepositContract that just stores
 ///        all funds without doing anything with them.
 ///
 ///        Should be able to work with proxy contracts so the contract
 ///        can be updated easily (but with great caution and transparency!)
 ///        when necessary.
 ///
 /// @author Brecht Devos - <brecht@loopring.org>
-contract BasicDepositContract is IDepositContract, ReentrancyGuard
+contract BasicDepositContract is IDepositContract, ReentrancyGuard, Claimable
 {
     using AddressUtil       for address;
     using ERC20SafeTransfer for address;
     using MathUint          for uint;
 
-    event TokenNotOwnedByUsersWithdrawn(
-        address sender,
-        address token,
-        address feeVault,
-        uint    amount
-    );
-
-    // Index base
-    uint constant public INDEX_BASE = 10 ** 18;
-
     address public exchange;
-    ILoopringV3 public loopring;
 
-    // A map from token address to the total exchange balances
-    mapping (address => uint) public exchangeBalance;
+    mapping (address => bool) needCheckBalance;
 
-    modifier onlyWhenUninitialized()
+    modifier onlyExchange()
     {
-        require(exchange == address(0), "INITIALIZED");
+        require(msg.sender == exchange, "UNAUTHORIZED");
         _;
     }
 
-    modifier onlyExchange()
+    modifier ifNotZero(uint amount)
     {
-        require(msg.sender == exchange, "UNAUTHORIZED");
-        _;
+        if (amount == 0) return;
+        else { _; }
     }
 
+    event CheckBalance(
+        address indexed token,
+        bool            checkBalance
+    );
+
+    event Deposit   (address token, uint amountReceived);
+    event Withdrawal(address token, uint amountPaid);
```

Comment:
We would now throw deposit/withdrawal events here and in the exchange contract, seems unnecessary to throw them on both contracts.
The exchange contract is better fit for the withdrawal events because it can also throw an event when a withdrawal isn't successful (or we could log successful withdrawals in the deposit contract and unsuccessful withdrawals on the exchange contract... but that seems a bit confusing).

Reply:
Agreed, removed.